torch/csrc/dynamo/eval_frame.c

```diff
@@ -1017,7 +1102,25 @@ static PyObject* _custom_eval_frame(
     // extract_cache_entry returns a borrowed reference. Modifying a borrowed
     // reference seems wrong. Therefore, we directly access the
     // extra->cache_entry. extra wont be NULL here.
-    extra->cache_entry = create_cache_entry(extra->cache_entry, result);
+    extra->cache_entry = create_cache_entry(extra->cache_entry, result, extra);
+
+    PyObject* check_fn = PyObject_GetAttrString(result, "check_fn");
+    DEBUG_NULL_CHECK(check_fn);
+
+    #ifdef TORCHDYNAMO_DEBUG
+    PyObject* check_fn_cache_entry = PyObject_GetAttrString(check_fn, "cache_entry");
+    DEBUG_CHECK(check_fn_cache_entry == Py_None);
+    Py_XDECREF(check_fn_cache_entry);
+    #endif
+
+    // Use a weakref to avoid circular references
```

Comment:
Add a comment on how this would be used to setup finalizer in guards.py, point to function `invalidate`